Allow constructing MediaStreamTrackProcessor from MediaStreamTrackHandle

youennf opened the pull request:
#### f03c4748b44a49eb535c200a030e3b199889d503
<pre>
Allow constructing MediaStreamTrackProcessor from MediaStreamTrackHandle
<a href="https://rdar.apple.com/167029476">rdar://167029476</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=304596">https://bugs.webkit.org/show_bug.cgi?id=304596</a>

Reviewed by Eric Carlson.

Test: http/wpt/mediastream/mediastreamtrackprocessor-mediastreamtrackhandle.html
* LayoutTests/http/wpt/mediastream/mediastreamtrackprocessor-mediastreamtrackhandle-expected.txt: Added.
* LayoutTests/http/wpt/mediastream/mediastreamtrackprocessor-mediastreamtrackhandle.html: Added.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::keeper):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
(WebCore::MediaStreamTrack::Keeper::create):
(WebCore::MediaStreamTrack::Keeper::isTrackEnabled const):
(WebCore::MediaStreamTrack::Keeper::setEnabled):
(WebCore::MediaStreamTrack::Keeper::Keeper):
* Source/WebCore/Modules/mediastream/MediaStreamTrackHandle.cpp:
(WebCore::MediaStreamTrackHandle::create):
(WebCore::MediaStreamTrackHandle::MediaStreamTrackHandle):
(WebCore::MediaStreamTrackHandle::detach):
* Source/WebCore/Modules/mediastream/MediaStreamTrackHandle.h:
(WebCore::MediaStreamTrackHandle::trackContextIdentifier const):
(WebCore::MediaStreamTrackHandle::track const):
(WebCore::MediaStreamTrackHandle::protectedTrackSourceObserver const):
(WebCore::MediaStreamTrackHandle::trackKeeper const):
* Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.cpp:
(WebCore::MediaStreamTrackProcessor::create):
(WebCore::MediaStreamTrackProcessor::MediaStreamTrackProcessor):
(WebCore::m_trackObserver):
(WebCore::MediaStreamTrackProcessor::~MediaStreamTrackProcessor):
(WebCore::MediaStreamTrackProcessor::readable):
(WebCore::MediaStreamTrackProcessor::contextDestroyed):
(WebCore::MediaStreamTrackProcessor::stopObserving):
(WebCore::MediaStreamTrackProcessor::trackEnded):
(WebCore::MediaStreamTrackProcessor::Source::Source):
(WebCore::MediaStreamTrackProcessor::Source::trackEnded):
(WebCore::MediaStreamTrackProcessor::Source::doPull):
(WebCore::MediaStreamTrackProcessor::Source::doCancel):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::create):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::TrackObserverWrapper):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::start):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::stop):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::trackEnded):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::removeObserver):
(WebCore::m_track): Deleted.
(WebCore::MediaStreamTrackProcessor::stopVideoFrameObserver): Deleted.
(WebCore::MediaStreamTrackProcessor::Source::~Source): Deleted.
(WebCore::MediaStreamTrackProcessor::Source::close): Deleted.
* Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.h:
(WebCore::MediaStreamTrackProcessor::isEnabled const):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::~TrackObserverWrapper):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readMediaStreamTrackHandle):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivateSourceObserver::MediaStreamTrackPrivateSourceObserver):
(WebCore::MediaStreamTrackPrivateSourceObserver::initialize):
(WebCore::MediaStreamTrackPrivateSourceObserver::start):
(WebCore::MediaStreamTrackPrivateSourceObserver::stop):
(WebCore::MediaStreamTrackPrivateSourceObserver::requestToEnd):
(WebCore::MediaStreamTrackPrivateSourceObserver::setMuted):
(WebCore::MediaStreamTrackPrivateSourceObserver::close):
(WebCore::MediaStreamTrackPrivateSourceObserver::applyConstraints):
(WebCore::MediaStreamTrackPrivateSourceObserver::create): Deleted.
(WebCore::MediaStreamTrackPrivateSourceObserver::std::function&lt;void): Deleted.
(WebCore::MediaStreamTrackPrivateSourceObserver::source): Deleted.
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
(WebCore::MediaStreamTrackPrivateSourceObserver::create):
(WebCore::MediaStreamTrackPrivateSourceObserver::std::function&lt;void):
(WebCore::MediaStreamTrackPrivateSourceObserver::source):

Canonical link: <a href="https://commits.webkit.org/307071@main">https://commits.webkit.org/307071@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdb06b096f121d0cce65b7151f2fc864bf22b800

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/143306 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/15781 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/7137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/151982 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/96527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/f03d5e7f-676e-4c6f-a50f-5855bbdfaa4b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/145173 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/16441 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/15862 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/110215 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/96527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/bb2dbb07-a8df-48f5-9c30-bf644113de6e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/146255 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/12657 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/128633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/91124 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/1bd79302-d388-4cc0-8102-fa44e6924ecf) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/12141 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/9859 "Passed tests") | [✅ 🛠 wpe-libwebrtc](https://ews-build.webkit.org/#/builders/172/builds/1980 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/121569 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/167/builds/5098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/154293 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/15825 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/169/builds/6093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/118234 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/15862 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/13346 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/118574 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/30363 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/14512 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/126295 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/71220 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/15450 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/4928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/15184 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/79170 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/15395 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/15246 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.cpp:
```diff
@@ -47,53 +47,70 @@ WTF_MAKE_TZONE_ALLOCATED_IMPL(MediaStreamTrackProcessor);
-        return Exception { ExceptionCode::TypeError, "Track is not video"_s };
-
-    if (init.track->ended())
-        return Exception { ExceptionCode::TypeError, "Track is ended"_s };
```
Is it now legal to create a track processor from a track that has ended?

youennf replied:
I kept it for now, thanks.

eric-carlson commented on LayoutTests/http/wpt/mediastream/mediastreamtrackprocessor-mediastreamtrackhandle.html:
```diff
@@ -0,0 +1,135 @@
+    worker.postMessage({});
+    
+    const videoFrame2 = await new Promise(resolve => worker.onmessage = e => resolve(e.data));
+    test.add_cleanup(() => videoFrame1.close());
```
```suggestion
    test.add_cleanup(() => videoFrame2.close());
```

youennf committed:
Allow constructing MediaStreamTrackProcessor from MediaStreamTrackHandle
rdar://167029476
https://bugs.webkit.org/show_bug.cgi?id=304596

Reviewed by Eric Carlson.

Test: http/wpt/mediastream/mediastreamtrackprocessor-mediastreamtrackhandle.html
* LayoutTests/http/wpt/mediastream/mediastreamtrackprocessor-mediastreamtrackhandle-expected.txt: Added.
* LayoutTests/http/wpt/mediastream/mediastreamtrackprocessor-mediastreamtrackhandle.html: Added.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::keeper):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
(WebCore::MediaStreamTrack::Keeper::create):
(WebCore::MediaStreamTrack::Keeper::isTrackEnabled const):
(WebCore::MediaStreamTrack::Keeper::setEnabled):
(WebCore::MediaStreamTrack::Keeper::Keeper):
* Source/WebCore/Modules/mediastream/MediaStreamTrackHandle.cpp:
(WebCore::MediaStreamTrackHandle::create):
(WebCore::MediaStreamTrackHandle::MediaStreamTrackHandle):
(WebCore::MediaStreamTrackHandle::detach):
* Source/WebCore/Modules/mediastream/MediaStreamTrackHandle.h:
(WebCore::MediaStreamTrackHandle::trackContextIdentifier const):
(WebCore::MediaStreamTrackHandle::track const):
(WebCore::MediaStreamTrackHandle::protectedTrackSourceObserver const):
(WebCore::MediaStreamTrackHandle::trackKeeper const):
* Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.cpp:
(WebCore::MediaStreamTrackProcessor::create):
(WebCore::MediaStreamTrackProcessor::MediaStreamTrackProcessor):
(WebCore::m_trackObserver):
(WebCore::MediaStreamTrackProcessor::~MediaStreamTrackProcessor):
(WebCore::MediaStreamTrackProcessor::readable):
(WebCore::MediaStreamTrackProcessor::contextDestroyed):
(WebCore::MediaStreamTrackProcessor::stopObserving):
(WebCore::MediaStreamTrackProcessor::trackEnded):
(WebCore::MediaStreamTrackProcessor::Source::Source):
(WebCore::MediaStreamTrackProcessor::Source::trackEnded):
(WebCore::MediaStreamTrackProcessor::Source::doPull):
(WebCore::MediaStreamTrackProcessor::Source::doCancel):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::create):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::TrackObserverWrapper):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::start):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::stop):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::trackEnded):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::removeObserver):
(WebCore::m_track): Deleted.
(WebCore::MediaStreamTrackProcessor::stopVideoFrameObserver): Deleted.
(WebCore::MediaStreamTrackProcessor::Source::~Source): Deleted.
(WebCore::MediaStreamTrackProcessor::Source::close): Deleted.
* Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.h:
(WebCore::MediaStreamTrackProcessor::isEnabled const):
(WebCore::MediaStreamTrackProcessor::TrackObserverWrapper::~TrackObserverWrapper):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readMediaStreamTrackHandle):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivateSourceObserver::MediaStreamTrackPrivateSourceObserver):
(WebCore::MediaStreamTrackPrivateSourceObserver::initialize):
(WebCore::MediaStreamTrackPrivateSourceObserver::start):
(WebCore::MediaStreamTrackPrivateSourceObserver::stop):
(WebCore::MediaStreamTrackPrivateSourceObserver::requestToEnd):
(WebCore::MediaStreamTrackPrivateSourceObserver::setMuted):
(WebCore::MediaStreamTrackPrivateSourceObserver::close):
(WebCore::MediaStreamTrackPrivateSourceObserver::applyConstraints):
(WebCore::MediaStreamTrackPrivateSourceObserver::create): Deleted.
(WebCore::MediaStreamTrackPrivateSourceObserver::std::function<void): Deleted.
(WebCore::MediaStreamTrackPrivateSourceObserver::source): Deleted.
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
(WebCore::MediaStreamTrackPrivateSourceObserver::create):
(WebCore::MediaStreamTrackPrivateSourceObserver::std::function<void):
(WebCore::MediaStreamTrackPrivateSourceObserver::source):

Canonical link: https://commits.webkit.org/307071@main
